### ISSUE TITLE
receive_timeout no longer applies when Connection.start_tls is called

### ISSUE THREAD

**Issue by the-allanc** [cannatag/ldap3#356]
Took me quite a while to debug this. I'm on a 64 bit Ubuntu 16.04 install, using Python 3.5. I'm testing by establishing a VPN connection, opening an LDAP connection, disconnecting the VPN after some use, and then attempting to still interact with the same connection object.

If you set the `receive_timeout` parameter on a Connection object, it will no longer have an effect once you invoke `start_tls` - any interactions via the socket do not seem to timeout appropriately.

I believed this was because of the way the Python ssl library appears to work - wrapping a socket involves dismantling the original socket object and transporting attributes to a new SSLSocket instance (and presumably, any socket options set were not propagated across).

However, I've just tried to manually set the socket option to an SSLSocket object to restore the timeout (this is done after `start_tls` is invoked):
```
conn.socket.setsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO, pack('LL', 10, 0))
```

But doing this hasn't fixed the issue - if I then try to use the connection, it still hangs rather than timing out after 10 seconds.

I'm not sure whether this is an issue with how Python implements SSL, or an issue with ldap3, or whether this is just something that is incompatible with SSL and needs documenting.

**Comment by cannatag** [comment]
HI, I'm not able to replicate your issue. In my testing even when I use a ssl socket with the start_tls operation the receive timeout is still valid.

Can you try the Timeouts_test.py script in the root of the github ldap3 project? You have just to set the SERVER variable to your connection string. The script will try various operation, requesting you to make the server UNAVAILABLE or REACHABLE. Then it writes two files timeout_test_ldap3.txt and timeout_test_ldap3.log. Can you send me both?

The script is quite simple. Feel free to adapt it to your testing needs to replicate your problem.

Thanks,
Giovanni

**Comment by the-allanc** [comment]
I'm including the two log files as requested - it behaves well until TLS is used, and then the 3 second timeout ends up being 1076 seconds instead. I've asked someone else on a different machine to see if they have the same behaviour using the same VPN and LDAP server.

[timeouts_test_ldap3.txt](https://github.com/cannatag/ldap3/files/1048713/timeouts_test_ldap3.txt)

[timeouts_test_ldap3.log.txt](https://github.com/cannatag/ldap3/files/1048714/timeouts_test_ldap3.log.txt)

**Comment by mattwbarry** [comment]
Here's my output from the test. Search timeout is only 148 seconds - less than 1076, but still way more than 3.


[timeouts_test_ldap3.log.txt](https://github.com/cannatag/ldap3/files/1048740/timeouts_test_ldap3.log.txt)
[timeouts_test_ldap3.txt](https://github.com/cannatag/ldap3/files/1048741/timeouts_test_ldap3.txt)

**Comment by cannatag** [comment]
Thanks, which ldap3 version are you using? Are you using a windows or a Linux box? What is your operating system?

**Comment by mattwbarry** [comment]
ldap3 - 2.2.4
OSX - 10.12.4
Python - 3.6.1

**Comment by cannatag** [comment]
On WIndows the script is working as expected:

```
CONNECT TIMEOUT 2 - RECEIVE TIMEOUT 3
connection timeout    : 000.000986
search timeout        : 003.004356
search timeout (TLS)  : 003.00559
```
On Linux I have the same problem. Probably the ssl wrapping is different between WIndows and Linux. Will investigate.

**Comment by the-allanc** [comment]
Just a couple of follow up comments:

I adapted the script slightly to get it to work with Python 2, but I still had the same experience as I did with Python 3.

I added the following line before and after `c.start_tls()`: `print(repr(c.socket), 'has sockopt', repr(c.socket.getsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO)))`

And it printed the following:
```<socket.socket fd=5, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.32.21.14', 32964), raddr=('10.32.4.240', 389)> has sockopt 3```
```<ssl.SSLSocket fd=5, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.32.21.14', 32964), raddr=('10.32.4.240', 389)> has sockopt 3```

So the socket option still seems to be applied once the socket is converted to an SSL socket.

**Comment by the-allanc** [comment]
I've come up with a test which seems to imply this is a Python issue, rather than with this library. I've attached a script which makes both a HTTP and HTTPS connection to the httpbin website, and both which set a socket read timeout in the same way that ldap3 does.

The HTTP test will timeout correctly. The HTTPS one will not. I think I'll file this as a Python issue.

[httpdelay.py.zip](https://github.com/cannatag/ldap3/files/1049227/httpdelay.py.zip)

**Comment by cannatag** [comment]
Thanks for your testing. It has been very helpful. Can you try now the ldap3 code in dev? You could try again the timeouts_test.py script with the latest dev code and tell me if now it's working now,.

**Comment by cannatag** [comment]
Any news? I'm going to release the 2.2.5 version soon and I'd like to know if it worked for you.

**Comment by cannatag** [comment]
Fixed in release 2.3 of ldap3.
